shell/platform/common/client_wrapper/include/flutter/event_channel.h

```diff
@@ -61,11 +64,12 @@ class EventChannel {
     const MethodCodec<T>* codec = codec_;
     const std::string channel_name = name_;
     const BinaryMessenger* messenger = messenger_;
+    bool is_listening = false;
```

Comment:
Instead of mutating a captured variable, don't you think it would be more clear to capture a reference to a bool on the heap?

Reply:
I didn't mean literally a C++ reference obviously, a `std::unique_ptr<bool>` just to be clear =)

Reply:
I agree mutating a captured variable is very subtle. It looks like I'll need to use `std::shared_ptr` per this comment:

https://github.com/flutter/engine/blob/d0791965483dee6f98aa73f81fd481944a182c5e/shell/platform/common/client_wrapper/include/flutter/event_channel.h#L57-L60

Reply:
Updated using `shared_ptr`, let me know if you have additional feedback! :)

Reply:
Did the mutable lands not end up working then?

Reply:
Oh, bummer about shared_ptr.  Weird it has to be copyable, I suspect that isn't a real requirement.  This looks good to me still.

Reply:
@stuartmorgan mutable lambda did work, we just felt mutating the `is_listening` variable was very subtle:

https://github.com/flutter/engine/blob/4bbf1ce7e0fff8cbd7b19d60845c7ab16a528419/shell/platform/common/client_wrapper/include/flutter/event_channel.h#L67-L130

Reply:
> we just felt mutating the `is_listening` variable was very subtle

Fair enough, although I don't think you actually had to capture anything. IIRC you can declare the variable inline in the capture list, which makes it much clearer that it's lamba state.

(I'm fine with this version as well, but it seems like you're basically hand-rolling a version of mutable lambda here.)

> I suspect that isn't a real requirement.

As the author of the comment, I assure you it is. Feel free to try changing that to a `unique_ptr` locally if you want to see the resulting compiler error.

> The intention is there is one bool for this one closure, this makes that happen. Copying and mutating the captured bool had N bools.

There should only ever conceptually be one copy of this lambda, even though it's technically copyable, so in practice N=1.

Reply:
> As the author of the comment, I assure you it is. Feel free to try changing that to a unique_ptr locally if you want to see the resulting compiler error.

Right, I believe that, but I think that's only the case because we are failing to `std::move` the closure correctly.  The comment states that there should only be one at a time, so it probably can be made to use `std::move` and then we could capture the `std::unique_ptr`.

Reply:
From my understanding, the issue isn't that we're failing to `std::move` the closure correctly (I couldn't find any missing `std::move`s), but rather that `std::function` requires that the wrapped lambda is copy-constructible. Capturing a `std::unique_ptr` makes it not copy-constructible.

These were the StackOverflow articles I found on the topic:
* https://stackoverflow.com/questions/58385454/how-to-return-a-lambda-with-a-captured-unique-ptr
* https://stackoverflow.com/questions/8236521/how-to-capture-a-unique-ptr-into-a-lambda-expression

Reply:
I also think it would be clearest to declare `is_listening` within the lambda captures as suggested in https://github.com/flutter/engine/pull/36882#discussion_r1000918746
That would represent how `is_listening` is mutable state that is local to the lambda.

The engine also provides an `fml::MakeCopyable` helper that wraps lambdas which capture move-only types like `std::unique_ptr`.  With `fml::MakeCopyable` you can avoid wrapping `handler` in a `shared_ptr`.

So this would look like:
```
    BinaryMessageHandler binary_handler = fml::MakeCopyable(
        [handler = std::move(handler),
         codec = codec_,
         channel_name = name_,
         messenger = messenger_,
         is_listening = bool(false)] (
           const uint8_t* message,
           const size_t message_size,
           BinaryReply reply) mutable {
```

Reply:
I mean, another way to look at it, one captured variable that is explicitly on the heap and mutable is easier to maintain than all the captures being mutable *shrug*.

Reply:
With `fml::MakeCopyable` there will only be one instance of the bool.  

`MakeCopyable` wraps a single instance of the lambda in a reference-counted object.

Reply:
The client wrapper code is external to the engine; it's an artifact that is copied into projects, to provide a more usable API than the C API we have to provide at the library level. So we're can't use FML, we'd have to duplicate that code into the wrapper if we wanted to use it.

Reply:
Here's a more simple and portable implementation of MakeCopyable:
```c++
#include <functional>
#include <iostream>

namespace {
  template <typename T>
  class MakeCopyable {
  public:
    MakeCopyable(T&& val) : ptr_(std::make_shared<T>(std::move(val))) {}

    template <typename... ArgType>
    auto operator()(ArgType&&... args) const {
      return ptr_->operator()(std::forward<ArgType>(args)...);
    }
   private:
    std::shared_ptr<T> ptr_;
  };
}

void RunTwice(const std::function<void()>& func) {
  func();
  func();
}

int main() {
  int32_t adder = 10;
  RunTwice(MakeCopyable([adder, tally = std::make_unique<int32_t>(0)]() {
    std::cout << adder + *tally << std::endl;
    *tally += 1;
  }));
}
```

Reply:
Thanks Jason for the info on `fml::MakeCopyable`, I wasn't aware of that!

Like Stuart mentioned, the client wrapper is code that's copied into the user's project, which the user can freely reference. I lean towards being conservative in adding new types to the client wrapper (though I realize `MakeCopyable` could likely be hidden as an implementation detail).

It seems folks are leaning more towards the mutable lambda now, I'll switch back to that 😄

Reply:
> though I realize `MakeCopyable` could likely be hidden as an implementation detail

It's non-trivial to hide most things in the wrapper since it's heavily templated.